### PR TITLE
build: update operator-sdk to v1.0

### LINKED ISSUE #678: Migrate to Operator SDK 1.0
We need to migrate the operator to use version 1.0 of the SDK. This will bring a number of benefits, not least of which is using a “supported” stable SDK: better kubebuilder integration (so we can generate RBAC from code, in particular), flat bundle generation (for the CSVs)...

### PR DESCRIPTION
**Pull request by SteveMattar**
### New installed tools:
- controller-gen generates RBAC manifests and crds
- kustomize manages declarative configuration for each content using kustomization.yaml

### Kubebuilder-style project:
- move the CRDs from deploy/crds/ to config/crd/bases
- move the CRs from deploy/crds/ to config/samples
- move the Controller deploy/operator.yaml to config/manager/manager.yaml
- move the RBAC manifests from deploy to config/rbac/
- move pkg/apis to apis
- move pkg/controller to controllers
- move build/Dockerfile to the project's root directory
- move cmd/manager/main.go to the project's root directory
- preparations for kustomize-based deployment for prometheus, webhooks and cert-manager
- scorecard configuration to test and validate the bundle
- replace the `+k8s:deepcopy-gen:interfaces=` marker with `+kubebuilder:object:root=true` for each api type
- add the marker `+operator-sdk:csv:customresourcedefinitions` for each api
- rename each Reconciler from `Reconcile<kind>` to `<kind>Reconciler`
- replace the `Add` function with `SetupWithManager`
- operator-lib module was added in order to keep the SDK leader.Become
- save the bundle and packagemanifests for the current version
- no need for zz_generated files anymore!
- OPERATOR_NAME and POD_NAME environment variable are no longer used
- envtest framework is not yet implemented

### Add More targets to Makefile:
- `make generate` handles deepcopy and adds the license text
- `make manifests` uses controller-gen and replaces the old commands `operator-sdk generate k8s`, `operator-sdk generate openapi` and `operator-sdk generate crds`

### Dapper:
- install controller-gen and kustomize

### Other changes:
- rename the globalnet ServiceAccount to follow [RFC1123](https://tools.ietf.org/html/rfc1123)
- update the CRs
- add more entries to yamllint
- handle the embedded yamls and support the new directory structure when generating crds

### TODO:

- [ ] main.go zap.FlagSet() is no longer supported
- [ ] main.go in opeartor-sdk v1 some utilities were moved to internal
- [x] automation for adding the fields displayName and description to customresourcedefinitions.owned[]

#### build the operator docker image
- make operator-image

#### bundle format
- make bundle VERSION=0.6.1 CHANNELS=alpha DEFAULT_CHANNEL=alpha
- make bundle-build

#### packagemanifests format
- make packagemanifests VERSION=0.6.1 FROM_VERSION=0.5.0 CHANNEL=alpha

### Refs:
- [migration guide](https://sdk.operatorframework.io/docs/building-operators/golang/project_migration_guide)
- [controller-gen](https://book.kubebuilder.io/reference/controller-gen.html)
- [kustomize](https://kubectl.docs.kubernetes.io/pages/reference/kustomize.html)

**BREAKING CHANGE**: migrating the legacy project to the new layout. closes #678

Signed-off-by: Steve Mattar <smattar@redhat.com>

### CODE REVIEW

File: Dockerfile.dapper
Comment (mkolesnik) [inline Dockerfile.dapper:9]: Defining `KUSTOMIZE` and `CONTROLLER_GEN` here without using them in the Dockerfile strikes me as odd.

Furthermore, we're not doing this "style" for any other command so why are these two special?

Finally, /go/bin is already a part of `$PATH` so you don't actually need to specify it.

Comment (SteveMattar) [reply Dockerfile.dapper:9]: Thanks for the review @mkolesnik I will remove those envs and call the cli commands directly in `Makefile`

**Comment by SteveMattar** [comment]
@mangelajo my e2e test fails because the new version of controller-runtime registers the flag `--kubeconfig` and it conflicts with ours the one in `subctl` command
https://github.com/kubernetes-sigs/controller-runtime/blob/c000ea850121b53dd769f624f3bfa74a4fcf100f/pkg/client/config/config.go#L38
do you have any experience with this? is there any workaround?

File: hack/boilerplate.go.txt
Comment (skitt) [inline hack/boilerplate.go.txt:2]: Please add a copyright header here:

```suggestion
© 2020 Red Hat, Inc. and others.
```

**Review by skitt — CHANGES_REQUESTED** [review]
A few initial comments, I’m looking into why this fails e2e.

Please also rebase this on top of current master, handling the conflicts, and ideally, squash the commits which fix issues in earlier commits with those commits (so the commits are never broken).

File: Makefile
Comment (SteveMattar) [reply Makefile:88]: you mean `make package/.image.submariner-operator` ? I think we can use this target I'll try it locally first

File: Makefile
Comment (SteveMattar) [reply Makefile:126]: yes, you are right I will revert this change

File: hack/boilerplate.go.txt
Comment (SteveMattar) [reply hack/boilerplate.go.txt:2]: `controller-gen` has a `year` argument I can use it to automate the process

File: .gitignore
Comment (dfarrell07) [inline .gitignore:15]: Running many of the make targets locally, removing this .gitignore change, I don't see any modified files.

Imho, we should recommend folks configure their editors/IDEs to store their files in some location other than the repo. Generally, I think we should try to avoid getting into the business of ignores rules for dev-specific tools.

File: Makefile
Comment (dfarrell07) [inline Makefile:165]: Does this need to be included in the release process somehow? I ask because when I run `make kustomization`, I get a modified version of `config/manager/kustomization.yaml`.

```
-  newTag: 0.7.0
+  newTag: v0.7.0-16-g80b6c81-dev
```

**Review by dfarrell07 — CHANGES_REQUESTED** [review]
Thanks for all the work @SteveMattar! Some comments inline.

Also, I don't know how to comment inline on commit messages, but it would be ideal to start body sentences with Caps and end with periods. See: https://submariner.io/contributing/code-review/#commit-message-formatting and https://chris.beams.io/posts/git-commit/.

File: Makefile
Comment (dfarrell07) [reply Makefile:175]: Maybe we should think about making CI cover this somehow?

**Comment by SteveMattar** [comment]
Thanks for your time @dfarrell07 
I will edit the commit messages

File: Makefile
Comment (SteveMattar) [reply Makefile:43]: ACK I will change that

File: Makefile
Comment (SteveMattar) [reply Makefile:175]: @dfarrell07 it fails because the calculated version includes some invalid characters for the bundle
you should specify the `VERSION` variable before running `make bundle`. kindly see the PR description for an example.

File: Makefile
Comment (SteveMattar) [reply Makefile:187]: same here as done in `make bundle`

File: .gitignore
Comment (SteveMattar) [reply .gitignore:15]: I agree I'll remove this section

File: Makefile
Comment (SteveMattar) [reply Makefile:165]: do you mean the new release repo? then yes.
`make bundle` and `make packagemanifests` both create a new bundle in different formats
and both use `make kustomization` as a preparation step

File: .gitignore
Comment (SteveMattar) [reply .gitignore:15]: I agree, I'll remove this section

**Review by SteveMattar — COMMENTED** [review]
Thanks for your time @dfarrell07
I will edit the commit messages

File: Makefile
Comment (dfarrell07) [reply Makefile:175]: Talking about this with @SteveMattar on the Automation WG sync today, the plan is to try to validate the version (if easy, ie doesn't have "-") and fail more quickly/clearly.

Imho the ideal is to have targets that pass by default (without any var overrides), but quickly and clearly failing is also okay.

File: Makefile
Comment (dfarrell07) [reply Makefile:165]: Thanks @SteveMattar, that makes sense. :+1: 

Are there any updates we need to make to https://submariner.io/contributing/release-process/?

Comment (SteveMattar) [reply Makefile:165]: @dfarrell07 yes in [step 9](https://submariner.io/contributing/release-process/#step-9-update-the-submariner-operator-on-operatorhubio) I have already prepared a commit locally for the submariner-website but waiting for this one to be merged first

File: Makefile
Comment (SteveMattar) [reply Makefile:182]: following our discussion I've added a validation using regex

File: Makefile
Comment (dfarrell07) [reply Makefile]: Imho it would be best to go ahead and push it, mark as draft, and we can start reviewing but hold on merging until this lands.

**Review by dfarrell07 — CHANGES_REQUESTED** [review]

File: Makefile
Comment (SteveMattar) [reply Makefile]: ACK

File: Makefile
Comment (SteveMattar) [reply Makefile:169]: Not exactly, kubebuilder is just the framework. The `operator-sdk` tool which generates the bundle fails as it expects a semantic version. I understand what you are saying, what do you think of this message?
```suggestion
	(printf '\nerror: Invalid VERSION - Please supply a semantic version for your bundle or package.\n\n' && exit 1)
```

**Comment by skitt** [comment]
@SteveMattar could you rebase this and resolve the conflicts? @dfarrell07 could you see if your comments have been addressed?

Thanks!

File: Makefile
Comment (dfarrell07) [reply Makefile]: I guess what I'm trying to clarify is that a version "major.minor.patch-pre" is still a valid semantic version, per https://github.com/submariner-io/submariner-operator/pull/786.

**Review by dfarrell07 — CHANGES_REQUESTED** [review]
When I run

```
make bundle VERSION=0.7.0 CHANNELS=alpha DEFAULT_CHANNEL=alpha
```

I see

```
--- a/bundle/manifests/submariner.io_submariners.yaml
+++ b/bundle/manifests/submariner.io_submariners.yaml
@@ -473,6 +473,8 @@ spec:
               type: object
             natEnabled:
               type: boolean
+            networkPlugin:
+              type: string
             routeAgentDaemonSetStatus:
               properties:
                 lastResourceVersion:
diff --git a/config/crd/bases/submariner.io_submariners.yaml b/config/crd/bases/submariner.io_submariners.yaml
index c7a6aab..7d3bc58 100644
--- a/config/crd/bases/submariner.io_submariners.yaml
+++ b/config/crd/bases/submariner.io_submariners.yaml
@@ -539,6 +539,8 @@ spec:
               type: object
             natEnabled:
               type: boolean
+            networkPlugin:
+              type: string
             routeAgentDaemonSetStatus:
               properties:
                 lastResourceVersion:
```

It seems like we need manual updates after https://github.com/submariner-io/submariner-operator/pull/779.

**Comment by dfarrell07** [comment]
I see a lot of minor formatting changes (indentation mostly) when I run `make packagemanifests VERSION=0.7.0 FROM_VERSION=0.6.1 CHANNEL=alpha`. Is that something that should be updated?

**Comment by dfarrell07** [comment]
Regarding the PR message, I don't see a command `make docker-build`.

File: Makefile
Comment (SteveMattar) [reply Makefile]: @dfarrell07 you are right although in the operator-sdk [official documentation](https://sdk.operatorframework.io/docs/cli/operator-sdk_generate_bundle/) they use the same term.
I will use your suggestion and replace kubebuilder by operator-sdk.
Thanks!

**Comment by SteveMattar** [comment]
> When I run
> 
> ```
> make bundle VERSION=0.7.0 CHANNELS=alpha DEFAULT_CHANNEL=alpha
> ```
> 
> I see
> 
> ```
> --- a/bundle/manifests/submariner.io_submariners.yaml
> +++ b/bundle/manifests/submariner.io_submariners.yaml
> @@ -473,6 +473,8 @@ spec:
>                type: object
>              natEnabled:
>                type: boolean
> +            networkPlugin:
> +              type: string
>              routeAgentDaemonSetStatus:
>                properties:
>                  lastResourceVersion:
> diff --git a/config/crd/bases/submariner.io_submariners.yaml b/config/crd/bases/submariner.io_submariners.yaml
> index c7a6aab..7d3bc58 100644
> --- a/config/crd/bases/submariner.io_submariners.yaml
> +++ b/config/crd/bases/submariner.io_submariners.yaml
> @@ -539,6 +539,8 @@ spec:
>                type: object
>              natEnabled:
>                type: boolean
> +            networkPlugin:
> +              type: string
>              routeAgentDaemonSetStatus:
>                properties:
>                  lastResourceVersion:
> ```
> 
> It seems like we need manual updates after #779.

I rebased today. In my opinion these changes should be included in v0.8.0

**Comment by SteveMattar** [comment]
> Should we somehow handle the version replaced by a new bundle? When I run `make bundle VERSION=0.7.0 CHANNELS=alpha DEFAULT_CHANNEL=alpha`, I see:
> 
> ```
> --- a/bundle/manifests/submariner.clusterserviceversion.yaml
> +++ b/bundle/manifests/submariner.clusterserviceversion.yaml
>    maturity: alpha
>    provider:
>      name: submariner.io
> -  replaces: submariner.v0.6.1
> +  replaces: submariner.v0.0.0
> ```

@dfarrell07 if the previous version is smaller than the new version then everything should work well.
Otherwise, if you run the command more than once using the same `VERSION` then the value for `replaces` is reset. 
I tested it several times but let me know if you find any scenario which doesn't seem to make sense then I will handle this field manually.

**Comment by SteveMattar** [comment]
> I see a lot of minor formatting changes (indentation mostly) when I run `make packagemanifests VERSION=0.7.0 FROM_VERSION=0.6.1 CHANNEL=alpha`. Is that something that should be updated?

Yes, indentation and white space. Unfortunately, this is how these tools output the generated YAML files. I've added those files to the ignore list in `.yamllint`. Do you think we should use `yamlfmt` to automatically modify the offset.

**Comment by SteveMattar** [comment]
> Regarding the PR message, I don't see a command `make docker-build`.

@dfarrell07 I moved this operation to shipyard and forgot to update the PR message. Thanks for noticing :+1:

**Review by skitt — APPROVED** [review]
LGTM, let’s get this in and address any remaining issues in follow-up patches (nothing we rely on currently is broken).

**Comment by dfarrell07** [comment]
>  Do you think we should use yamlfmt to automatically modify the offset.

That's a good idea, but I also wonder if this is something we could help fix upstream.

> In my opinion these changes should be included in the next bundle v0.8.0

Yep, that makes sense.

Agreed, let's get this in and iterate on these minor details. Thanks for all the good work @SteveMattar!

**Review by dfarrell07 — APPROVED** [review]

**Pull request merged**
